corehq/form_processor/backends/sql/dbaccessors.py

```diff
@@ -873,28 +902,35 @@ def _attach_prefetch_models(objects_by_id, prefetched_models, link_field_name, c
         setattr(obj, cached_attrib_name, list(group))
 
 
-def _batch_iterate(batch_fn, next_start_from_fn, start_from=None, chunk_size=500):
+def _batch_iterate_over_shards(batch_fn, next_start_from_fn, start_from=None, chunk_size=500):
     """
     Iterate through a function in batches. Assumes the following signatures:
 
-    batch_fn(start_from, limit) - a function that returns sorted data in batches
+    batch_fn(start_from, limit=limit, using=db_alias) - a function that returns sorted data in batches
     next_start_from_fn(item) - a function that returns a "start_from" for a item for the next batch
     """
     start_from = start_from or datetime.min
-    # todo: this will greedily query the same data multiple times in a sharded setup. We should make it smarter
-    batch = batch_fn(start_from, limit=chunk_size)
-    while batch:
-        for item in batch:
-            yield item
-            next_start_from = next_start_from_fn(item)
-
-        if len(batch) == chunk_size:
-            # we got a full chunk so keep checking for more
-            assert next_start_from > start_from  # make sure we are making progress
-            start_from = next_start_from
-            batch = batch_fn(start_from, limit=chunk_size)
-        else:
-            batch = []  # equivalent to return
+    if not settings.USE_PARTITIONED_DATABASE:
+        db_list = [None]  # use the default database
+    else:
+        from corehq.sql_db.config import PartitionConfig
+        partition_config = PartitionConfig()
+        db_list = partition_config.get_form_processing_dbs()
+
+    for db_alias in db_list:
```

Comment:
Looking at this again I think I'll need to change it to get a batch from each DB instead of processing an entire DB at a time. Also the `start_from` var is going to get overridden